janus_server/src/binary_utils/job_driver.rs

```diff
@@ -0,0 +1,388 @@
+            // to avoid potentially timing out while waiting on _other_ jobs to finish being
+            // stepped. This is racy given that workers may complete (and relinquish their permits)
+            // concurrently with us acquiring jobs; but that's OK, since this can only make us
+            // underestimate the number of jobs we can acquire, and underestimation is acceptable
+            // (we'll pick up any additional jobs on the next iteration of this loop). We can't
+            // overestimate since this task is the only place that permits are acquired.
+
+            // TODO(brandon): only acquire jobs whose batch units have not already been collected (probably by modifying acquire_incomplete_aggregation_jobs)
+            let max_acquire_count = sem.available_permits();
+            info!(max_acquire_count, "Acquiring jobs");
+            let acquired_jobs = (self.incomplete_job_acquirer)(
+                self.datastore.clone(),
+                self.worker_lease_duration,
+                max_acquire_count,
+            )
+            .await;
+            let acquired_jobs = match acquired_jobs {
+                Ok(acquired_jobs) => acquired_jobs,
+                Err(err) => {
+                    error!(?err, "Couldn't acquire jobs");
+                    // Go ahead and step job discovery delay in this error case to ensure we don't
+                    // tightly loop running transactions that will fail without any delay.
+                    job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                    continue;
+                }
+            };
+            if acquired_jobs.is_empty() {
+                debug!("No jobs available");
+                job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                continue;
+            }
+            info!(acquired_job_count = acquired_jobs.len(), "Acquired jobs");
+
+            // Start up tasks for each acquired job.
+            job_discovery_delay = Duration::ZERO;
+            for (acquired_job, lease_expiry) in acquired_jobs {
+                task::spawn({
+                    // We acquire a semaphore in the job-discovery task rather than inside the new
+                    // job-stepper task to ensure that acquiring a permit does not race with
+                    // checking how many permits we have available in the next iteration of this
+                    // loop, to maintain the invariant that this task is the only place we acquire
+                    // permits.
+                    //
+                    // Unwrap safety: we have seen that at least `acquired_jobs.len()` permits are
+                    // available, and this task is the only task that acquires permits.
+                    let permit = Arc::clone(&sem).try_acquire_owned().unwrap();
+                    let this = Arc::clone(&self);
+                    let acquired_job_clone = acquired_job.clone();
```

Comment:
Stick this in an `Arc` or otherwise avoid cloning a structure of unknown size, IMO.

Reply:
Done, at the cost of some slightly ugly `Arc<AcquiredJob>` in various function signatures.